Birdee/Preprocessing.cpp

```diff
@@ -1844,7 +1844,43 @@ namespace Birdee
 				string("The function ") + curfunc->Proto->Name + " overrides the function at " + overriden->Pos.ToString()
 				+ " but they have different prototypes");
 		};
-    
+
+		// automatically mark class as abstract if there's unimplemented pure virtual function
+		{
+			unordered_set<string> parent_funcdef;
+			ClassAST* cls = parent_class;
+			while (!this->is_abstract && cls)
+			{
+				for (auto& funcdef : cls->funcs) {
+					string & name = funcdef.decl->Proto->Name;
+					if (parent_funcdef.find(name) != parent_funcdef.end())
+						continue;
+					parent_funcdef.insert(name);
+					if (funcdef.is_pure && this->funcmap.find(name) == this->funcmap.end()) {
+						this->is_abstract = true;
+						break;
+					}
+				}
+				cls = cls->parent_class;
+			}
+		}
+
+		for (auto& funcdef : funcs) {
+			if (funcdef.is_pure) {
+				ClassAST* cls = parent_class;
+				while (cls)
+				{
+					auto itr = cls->funcmap.find(funcdef.decl->Proto->Name);
+					if (itr != cls->funcmap.end())
+					{
+						CompileAssert(cls->funcs[itr->second].is_pure, funcdef.decl->Pos, "Cannot mark non-pure virtual function as pure");
```

Comment:
What does "Cannot mark non-pure virtual function as pure" means? Maybe a more specific prompt, like "method xxx is defined as pure virtual in class XXX, but it is non-virtual in parent class"